crates/bevy_animation_rig/src/skinned_mesh.rs

```diff
@@ -0,0 +1,369 @@
+use bevy_asset::{Assets, Handle, HandleUntyped};
+use bevy_ecs::{
+    entity::{Entity, EntityMap, MapEntities, MapEntitiesError},
+    reflect::{ReflectComponent, ReflectMapEntities},
+    system::{Query, Res, ResMut},
+};
+use bevy_math::Mat4;
+use bevy_pbr::render_graph;
+use bevy_reflect::{
+    serde, DynamicStruct, FieldIter, Reflect, ReflectMut, ReflectRef, Struct, TypeUuid,
+};
+use bevy_render::{
+    pipeline::PipelineDescriptor,
+    render_graph::{RenderGraph, RenderResourcesNode},
+    renderer::{
+        RenderResource, RenderResourceHints, RenderResourceIterator, RenderResourceType,
+        RenderResources,
+    },
+    shader::{Shader, ShaderStage},
+    texture::Texture,
+};
+use bevy_transform::components::GlobalTransform;
+
+/// The name of skinned mesh node
+pub mod node {
+    pub const SKINNED_MESH: &str = "skinned_mesh";
+}
+
+/// The name of skinned mesh buffer
+pub mod buffer {
+    pub const JOINT_TRANSFORMS: &str = "JointTransforms";
+}
+
+/// Specify RenderPipelines with this handle to render the skinned mesh.
+pub const SKINNED_MESH_PIPELINE_HANDLE: HandleUntyped =
+    HandleUntyped::weak_from_u64(PipelineDescriptor::TYPE_UUID, 0x14db1922328e7fcc);
+
+/// Used to update and bind joint transforms to the skinned mesh render pipeline specified with [`SKINNED_MESH_PIPELINE_HANDLE`].
+///
+/// The length of entities vector passed to [`SkinnedMesh::new()`] should equal to the number of matrices inside [`SkinnedMeshInverseBindposes`].
+///
+/// The content of `joint_transforms` can be modified manually if [`skinned_mesh_update`] system is disabled.
+///
+/// # Example
+/// ```
+/// use bevy_animation_rig::{SkinnedMesh, SKINNED_MESH_PIPELINE_HANDLE};
+/// use bevy_ecs::{entity::Entity, system::Commands};
+/// use bevy_pbr::prelude::PbrBundle;
+/// use bevy_render::pipeline::{RenderPipeline, RenderPipelines};
+///
+/// fn example_system(mut commands: Commands) {
+///     commands.spawn_bundle(PbrBundle {
+///         render_pipelines: RenderPipelines::from_pipelines(
+///             vec![RenderPipeline::new(SKINNED_MESH_PIPELINE_HANDLE.typed())]
+///         ),
+///         ..Default::default()
+///     }).insert(SkinnedMesh::new(
+///         // Refer to [`SkinnedMeshInverseBindposes`] example on how to create inverse bindposes data.
+///         Default::default(),
+///         // Specify joint entities here.
+///         vec![Entity::new(0)]
+///     ));
+/// }
+/// ```
+#[derive(Debug, Default, Clone, Reflect)]
+#[reflect(Component, MapEntities)]
+pub struct SkinnedMesh {
+    pub inverse_bindposes: Handle<SkinnedMeshInverseBindposes>,
+    pub joints: Vec<SkinnedMeshJoint>,
+}
+
+impl SkinnedMesh {
+    pub fn new(
+        inverse_bindposes: Handle<SkinnedMeshInverseBindposes>,
+        joint_entities: Vec<Entity>,
+    ) -> Self {
+        Self {
+            inverse_bindposes,
+            joints: joint_entities
+                .iter()
```

Comment:
```suggestion
                .into_iter()
```
We should probably call `into_iter` here to better express that we're consuming the incoming vector.